components/script/dom/servoparser/prefetch.rs

```diff
@@ -0,0 +1,71 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#![allow(unrooted_must_root)]
+
+use html5ever::buffer_queue::BufferQueue;
+use html5ever::tokenizer::Token;
+use html5ever::tokenizer::TokenSink;
+use html5ever::tokenizer::TokenSinkResult;
+use html5ever::tokenizer::Tokenizer as HtmlTokenizer;
+use net_traits::request::RequestBuilder;
+use net_traits::CoreResourceMsg;
+use net_traits::FetchChannels;
+use net_traits::IpcSend;
+use net_traits::ResourceThreads;
+use servo_url::ServoUrl;
+
+#[derive(MallocSizeOf)]
+#[must_root]
+pub struct Tokenizer {
+    #[ignore_malloc_size_of = "Defined in html5ever"]
+    inner: HtmlTokenizer<PrefetchSink>,
+}
+
+unsafe_no_jsmanaged_fields!(Tokenizer);
```

Comment:
I would prefer to derive JSTraceable on Tokenizer instead to avoid accidents in the future if we update that struct.

Reply:
Oh this is weird, I replied to this comment but github appears to have lost the reply...

Anyway, the problem is that HtmlTokenizer doesn't implement `JSTraceable`, and it's difficult to do so as its generic, and the tokenizer API provides no way to access the sink. If it did, the implementation would be pretty easy:
```rust
    impl<Sink: JSTraceable> for HtmlTokenizer<Sink> {
        fn trace(&self) { self.sink().trace() } 
   } 
```
but there's no `sink()` method, so we can't really do that.

Reply:
We have a JSTraceable implementation for the HTML tokenizer in script/dom/servoparser/html.rs. I recommend following its example.

Reply:
Er, but stripping out the unnecessary code leaves just:
```rust
#[allow(unsafe_code)]
unsafe impl JSTraceable for HtmlTokenizer<PrefetchSink>> {
    unsafe fn trace(&self, trc: *mut JSTracer) {
    }
}
```
which doesn't have obvious advantages over `unsafe_no_jsmanaged_fields!`.

Reply:
Can't we still derive JStraceable on Sink and call `self.sink.trace()` at least?

Reply:
The problem is we don't have access to the sink, since the tokenizer doesn't expose an API to access it. We couuuld use an Rc, but that seems a bit round-the-houses.

Reply:
Isn't the tokenizer [this type](https://docs.rs/html5ever/0.23.0/html5ever/tokenizer/struct.Tokenizer.html). It has a public `sink` field. What am I missing?

Reply:
Argh, I was looking for an accessor method, I didn't spot that the field was public. D'oh!

Reply:
Okay, fixed.